# **ARTIFICIAL NEURAL NETWORKS**

## **1. Data Exploration and Preprocessing**

In [43]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as snns
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler

In [44]:
df = pd.read_csv('/content/Alphabets_data.csv')
df

letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  \
0          T     2     8      3       5      1     8    13      0      6   
1          I     5    12      3       7      2    10     5      5      4   
2          D     4    11      6       8      6    10     6      2      6   
3          N     7    11      6       6      3     5     9      4      6   
4          G     2     1      3       1      1     8     6      6      6   
...      ...   ...   ...    ...     ...    ...   ...   ...    ...    ...   
19995      D     2     2      3       3      2     7     7      7      6   
19996      C     7    10      8       8      4     4     8      6      9   
19997      T     6     9      6       7      5     6    11      3      7   
19998      S     2     3      4       2      1     8     7      2      6   
19999      A     4     9      6       6      2     9     5      3      1   

       xybar  x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0          6      10       8      0       8      0       8  
1         13       3       9      2       8      4      10  
2         10       3       7      3       7      3       9  
3          4       4      10      6      10      2       8  
4          6       5       9      1       7      5      10  
...      ...     ...     ...    ...     ...    ...     ...  
19995      6       6       4      2       8      3       7  
19996     12       9      13      2       9      3       7  
19997     11       9       5      2      12      2       4  
19998     10       6       8      1       9      5       8  
19999      8       1       8      2       7      2       8  

[20000 rows x 17 columns]

In [45]:
df.describe()

xbox          ybox         width       height         onpix  \
count  20000.000000  20000.000000  20000.000000  20000.00000  20000.000000   
mean       4.023550      7.035500      5.121850      5.37245      3.505850   
std        1.913212      3.304555      2.014573      2.26139      2.190458   
min        0.000000      0.000000      0.000000      0.00000      0.000000   
25%        3.000000      5.000000      4.000000      4.00000      2.000000   
50%        4.000000      7.000000      5.000000      6.00000      3.000000   
75%        5.000000      9.000000      6.000000      7.00000      5.000000   
max       15.000000     15.000000     15.000000     15.00000     15.000000   

               xbar          ybar         x2bar         y2bar         xybar  \
count  20000.000000  20000.000000  20000.000000  20000.000000  20000.000000   
mean       6.897600      7.500450      4.628600      5.178650      8.282050   
std        2.026035      2.325354      2.699968      2.380823      2.488475   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        6.000000      6.000000      3.000000      4.000000      7.000000   
50%        7.000000      7.000000      4.000000      5.000000      8.000000   
75%        8.000000      9.000000      6.000000      7.000000     10.000000   
max       15.000000     15.000000     15.000000     15.000000     15.000000   

            x2ybar        xy2bar         xedge        xedgey         yedge  \
count  20000.00000  20000.000000  20000.000000  20000.000000  20000.000000   
mean       6.45400      7.929000      3.046100      8.338850      3.691750   
std        2.63107      2.080619      2.332541      1.546722      2.567073   
min        0.00000      0.000000      0.000000      0.000000      0.000000   
25%        5.00000      7.000000      1.000000      8.000000      2.000000   
50%        6.00000      8.000000      3.000000      8.000000      3.000000   
75%        8.00000      9.000000      4.000000      9.000000      5.000000   
max       15.00000     15.000000     15.000000     15.000000     15.000000   

            yedgex  
count  20000.00000  
mean       7.80120  
std        1.61747  
min        0.00000  
25%        7.00000  
50%        8.00000  
75%        9.00000  
max       15.00000

In [46]:
df.dtypes

letter    object
xbox       int64
ybox       int64
width      int64
height     int64
onpix      int64
xbar       int64
ybar       int64
x2bar      int64
y2bar      int64
xybar      int64
x2ybar     int64
xy2bar     int64
xedge      int64
xedgey     int64
yedge      int64
yedgex     int64
dtype: object

In [47]:
df.isnull().sum()

letter    0
xbox      0
ybox      0
width     0
height    0
onpix     0
xbar      0
ybar      0
x2bar     0
y2bar     0
xybar     0
x2ybar    0
xy2bar    0
xedge     0
xedgey    0
yedge     0
yedgex    0
dtype: int64

In [48]:
# Separating features and target
features = df.drop('letter', axis=1)
target = df['letter']

In [49]:
# Normalizing the feature columns
#scaler = StandardScaler()
#features_scaled = scaler.fit_transform(features)

In [50]:
scaler = MinMaxScaler()
features_scaled = scaler.fit_transform(features)

In [51]:
# Converting the scaled features back to a DataFrame
df_scaled = pd.DataFrame(features_scaled, columns=features.columns)

In [52]:
# Appending the target column back to the scaled DataFrame
df_scaled['letter'] = target

In [53]:
# Displaying the first few rows of the processed dataset
print(df_scaled.head())

       xbox      ybox  width    height     onpix      xbar      ybar  \
0  0.133333  0.533333    0.2  0.333333  0.066667  0.533333  0.866667   
1  0.333333  0.800000    0.2  0.466667  0.133333  0.666667  0.333333   
2  0.266667  0.733333    0.4  0.533333  0.400000  0.666667  0.400000   
3  0.466667  0.733333    0.4  0.400000  0.200000  0.333333  0.600000   
4  0.133333  0.066667    0.2  0.066667  0.066667  0.533333  0.400000   

      x2bar     y2bar     xybar    x2ybar    xy2bar     xedge    xedgey  \
0  0.000000  0.400000  0.400000  0.666667  0.533333  0.000000  0.533333   
1  0.333333  0.266667  0.866667  0.200000  0.600000  0.133333  0.533333   
2  0.133333  0.400000  0.666667  0.200000  0.466667  0.200000  0.466667   
3  0.266667  0.400000  0.266667  0.266667  0.666667  0.400000  0.666667   
4  0.400000  0.400000  0.400000  0.333333  0.600000  0.066667  0.466667   

      yedge    yedgex letter  
0  0.000000  0.533333      T  
1  0.266667  0.666667      I  
2  0.200000  0.600000  

## **2. Model Implementation**

In [54]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, accuracy_score

In [55]:
# Convert target labels to one-hot encoding
y = pd.get_dummies(df_scaled['letter'])

In [56]:
# Split the data into features and target
X = df_scaled.drop('letter', axis=1)

In [57]:
# Step 2: Divide the Dataset into Training and Test Sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [58]:
# Step 3: Construct the ANN Model
model = Sequential()
model.add(Dense(64, input_shape=(X_train.shape[1],), activation='relu'))
model.add(Dense(32, activation='relu'))
model.add(Dense(y_train.shape[1], activation='softmax'))

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [59]:
# Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [60]:
# Step 4: Train the Model on the Training Set
history = model.fit(X_train, y_train, epochs=20, batch_size=32, validation_split=0.2)

Epoch 1/20
400/400 ━━━━━━━━━━━━━━━━━━━━ 3s 4ms/step - accuracy: 0.1166 - loss: 3.1039 - val_accuracy: 0.4544 - val_loss: 2.1312
Epoch 2/20
400/400 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.4760 - loss: 1.9153 - val_accuracy: 0.5594 - val_loss: 1.5862
Epoch 3/20
400/400 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.5779 - loss: 1.5150 - val_accuracy: 0.6056 - val_loss: 1.4155
Epoch 4/20
400/400 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6210 - loss: 1.3613 - val_accuracy: 0.6525 - val_loss: 1.3159
Epoch 5/20
400/400 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6353 - loss: 1.2837 - val_accuracy: 0.6669 - val_loss: 1.2391
Epoch 6/20
400/400 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6751 - loss: 1.1543 - val_accuracy: 0.6669 - val_loss: 1.1981
Epoch 7/20
400/400 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6725 - loss: 1.1561 - val_accuracy: 0.6969 - val_loss: 1.1359
Epoch 8/20
400/400 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6967 - loss: 1.0816 - val_accuracy: 0.

In [61]:
# Step 5: Make Predictions on the Test Set
y_pred = model.predict(X_test)
y_pred_classes = y_pred.argmax(axis=1)
y_test_classes = y_test.values.argmax(axis=1)

125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step


In [62]:
# Evaluate the model's performance
accuracy = accuracy_score(y_test_classes, y_pred_classes)
report = classification_report(y_test_classes, y_pred_classes, target_names=y.columns)

In [63]:
print(f"Accuracy: {accuracy}")
print(f"Classification Report:\n{report}")

Accuracy: 0.778
Classification Report:
              precision    recall  f1-score   support

           A       0.87      0.89      0.88       149
           B       0.66      0.83      0.74       153
           C       0.89      0.69      0.78       137
           D       0.64      0.83      0.73       156
           E       0.76      0.72      0.74       141
           F       0.74      0.78      0.76       140
           G       0.74      0.79      0.76       160
           H       0.76      0.56      0.64       144
           I       0.87      0.80      0.83       146
           J       0.75      0.86      0.80       149
           K       0.68      0.75      0.72       130
           L       0.84      0.80      0.82       155
           M       0.88      0.93      0.90       168
           N       0.89      0.79      0.84       151
           O       0.73      0.75      0.74       145
           P       0.87      0.81      0.84       173
           Q       0.83      0.69      0.7

## **3. Hyperparameter Tuning**

In [64]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report, accuracy_score
from sklearn.utils import class_weight

In [65]:
# Converting target labels to one-hot encoding
y = pd.get_dummies(df_scaled['letter'])

In [66]:
# Spliting the data into features and target
X = df_scaled.drop('letter', axis=1)

In [67]:
# Step 2: Divide the Dataset into Training and Test Sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [68]:
# Calculating class weights
class_weights = class_weight.compute_class_weight('balanced', classes=np.unique(y_train.values.argmax(axis=1)), y=y_train.values.argmax(axis=1))
class_weights_dict = dict(enumerate(class_weights))

In [69]:
# Defining the ANN model
model = Sequential()
model.add(Dense(256, input_shape=(X_train.shape[1],), activation='relu'))
model.add(Dense(128, activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dense(y_train.shape[1], activation='softmax'))

/usr/local/lib/python3.10/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [70]:
# Compiling the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [71]:
# Train the model with class weights
history = model.fit(X_train, y_train, epochs=50, batch_size=32, validation_split=0.2, class_weight=class_weights_dict)

Epoch 1/50
400/400 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.2541 - loss: 2.6426 - val_accuracy: 0.5884 - val_loss: 1.4052
Epoch 2/50
400/400 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.6321 - loss: 1.2755 - val_accuracy: 0.6816 - val_loss: 1.1422
Epoch 3/50
400/400 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6906 - loss: 1.0410 - val_accuracy: 0.7209 - val_loss: 0.9767
Epoch 4/50
400/400 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7390 - loss: 0.9060 - val_accuracy: 0.7650 - val_loss: 0.8370
Epoch 5/50
400/400 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7587 - loss: 0.8266 - val_accuracy: 0.7609 - val_loss: 0.7859
Epoch 6/50
400/400 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7695 - loss: 0.7575 - val_accuracy: 0.7828 - val_loss: 0.7071
Epoch 7/50
400/400 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7997 - loss: 0.6710 - val_accuracy: 0.8059 - val_loss: 0.6552
Epoch 8/50
400/400 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8125 - loss: 0.6329 - val_accuracy: 0.

In [72]:
# Displaying the model summary
model.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense_10 (Dense)                     │ (None, 256)                 │           4,352 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_11 (Dense)                     │ (None, 128)                 │          32,896 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_12 (Dense)                     │ (None, 64)                  │           8,256 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_13 (Dense)                     │ (None, 26)                  │           1,690 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 141,584 (553.07 KB)

 Trainable params: 47,194 (184.35 KB)

 Non-trainable params: 0 (0.00 B)

 Optimizer params: 94,390 (368.71 KB)

In [73]:
# Making Predictions on the Test Set
y_pred = model.predict(X_test)
y_pred_classes = y_pred.argmax(axis=1)
y_test_classes = y_test.values.argmax(axis=1)

125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step


In [74]:
# Evaluate the model's performance
accuracy = accuracy_score(y_test_classes, y_pred_classes)
report = classification_report(y_test_classes, y_pred_classes, target_names=y.columns)

In [75]:
print(f"Accuracy: {accuracy}")
print(f"Classification Report:\n{report}")

Accuracy: 0.9415
Classification Report:
              precision    recall  f1-score   support

           A       0.94      0.99      0.96       149
           B       0.97      0.89      0.93       153
           C       0.98      0.88      0.92       137
           D       0.97      0.95      0.96       156
           E       0.92      0.94      0.93       141
           F       0.94      0.91      0.92       140
           G       0.86      0.98      0.92       160
           H       0.90      0.89      0.89       144
           I       0.92      0.98      0.95       146
           J       1.00      0.91      0.95       149
           K       0.93      0.87      0.90       130
           L       0.93      0.97      0.95       155
           M       0.95      0.98      0.96       168
           N       0.99      0.86      0.92       151
           O       0.87      0.97      0.92       145
           P       0.99      0.92      0.96       173
           Q       0.96      0.99      0.

## **4. Evaluation**

In [76]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, classification_report

In [77]:
# Converting DataFrame to NumPy array and find the class with the highest probability
y_pred_classes = y_pred.argmax(axis=1)
y_test_classes = y_test.to_numpy().argmax(axis=1)

In [78]:
# Calculating evaluation metrics
accuracy = accuracy_score(y_test_classes, y_pred_classes)
precision = precision_score(y_test_classes, y_pred_classes, average='weighted')
recall = recall_score(y_test_classes, y_pred_classes, average='weighted')
f1 = f1_score(y_test_classes, y_pred_classes, average='weighted')

In [79]:
print(f"Accuracy: {accuracy}")
print(f"Precision: {precision}")
print(f"Recall: {recall}")
print(f"F1-Score: {f1}")
print("\nClassification Report:\n", classification_report(y_test_classes, y_pred_classes))

Accuracy: 0.9415
Precision: 0.9433788643946968
Recall: 0.9415
F1-Score: 0.9414801821307719

Classification Report:
               precision    recall  f1-score   support

           0       0.94      0.99      0.96       149
           1       0.97      0.89      0.93       153
           2       0.98      0.88      0.92       137
           3       0.97      0.95      0.96       156
           4       0.92      0.94      0.93       141
           5       0.94      0.91      0.92       140
           6       0.86      0.98      0.92       160
           7       0.90      0.89      0.89       144
           8       0.92      0.98      0.95       146
           9       1.00      0.91      0.95       149
          10       0.93      0.87      0.90       130
          11       0.93      0.97      0.95       155
          12       0.95      0.98      0.96       168
          13       0.99      0.86      0.92       151
          14       0.87      0.97      0.92       145
          15       

### **Discussion on Performance Differences**
**Performance Differences Between Default and Tuned Models:**  

1.   **Accuracy Improvement:**

*   Default Model: Often has lower accuracy due to the simplistic architecture and lack of optimization.
*   Tuned Model: Achieves higher accuracy (0.94 in this case) as a result of hyperparameter tuning and model optimization. This improvement suggests that tuning has significantly enhanced the model’s performance.

2.   **Precision, Recall, and F1-Score:**

*   Default Model: Typically shows varied performance across classes. Some classes may have very low precision or recall due to overfitting or underfitting.
*   Tuned Model: With tuning, precision, recall, and F1-scores improve substantially. The metrics are balanced across classes, indicating that the model has learned to generalize better and handle class imbalances effectively.

3.   **Generalization:**

*   Default Model: May overfit or underfit the training data due to a lack of complexity and optimization.
*   Tuned Model: Better generalizes to unseen data due to an optimized architecture and training process. This leads to consistent performance across various metrics.





